File: src/core/hasher.ts
/**
 * Content fingerprinting for detecting stable prompt segments across requests.
 *
 * Uses a fast string hash (djb2). We don't need cryptographic security here,
 * just collision-resistant identity checks for prompt content.
 */
export function hashContent(content: string): string {
  let hash = 5381
  for (let i = 0; i < content.length; i++) {
    hash = ((hash << 5) + hash + content.charCodeAt(i)) & 0xffffffff
  }
  return hash.toString(36)
}

export function hashObject(obj: unknown): string {
  return hashContent(JSON.stringify(obj))
}

/**
 * Hash a content block, normalizing whitespace so minor formatting
 * changes don't invalidate the cache.
 */
export function hashNormalized(content: string): string {
  const normalized = content.replace(/\s+/g, ' ').trim()
  return hashContent(normalized)
}

File: python/cachellm/core/hasher.py
import json
import re


def hash_content(content: str) -> str:
    """Content fingerprinting using djb2 hash.
    Not cryptographic — just collision-resistant identity checks for prompt content."""
    h = 5381
    for ch in content:
        h = ((h << 5) + h + ord(ch)) & 0xFFFFFFFF
    return _base36(h)


def hash_object(obj: object) -> str:
    return hash_content(json.dumps(obj, sort_keys=True))


def hash_normalized(content: str) -> str:
    """Hash content with normalized whitespace so minor formatting
    changes don't invalidate the cache."""
    normalized = re.sub(r"\s+", " ", content).strip()
    return hash_content(normalized)


def _base36(n: int) -> str:
    if n == 0:
        return "0"
    chars = "0123456789abcdefghijklmnopqrstuvwxyz"
    result = []
    while n:
        n, remainder = divmod(n, 36)
        result.append(chars[remainder])
    return "".join(reversed(result))
